### ISSUE TITLE
Consider calling a callback on event of removing a subscription

### ISSUE THREAD

**Issue by manifest** [vernemq/vernemq#1326]
When a client that use `clean_session=true` have gone, all its subscriptions get removed. The issue that there is no way to find out about that action from within plugin.

It seems as both `on_unsubscribe` and `on_unsubscribe_m5` are only called on explicit unsubscribe message from the client.  The function call `vmq_subscriber_db:read(SubscriberId)` also return an empty list at the moment of `on_client_gone` execution.

The proposal is:
- to call `on_client_gone` callback before subscriptions of the client were removed, so that information will be available through the call to `vmq_subscriber_db:read(SubscriberId)`
- or call `on_unsubscribe` and `on_unsubscribe_m5` on any subscription removal (not only explicit unsubscribe message of the client)
- or add a new callback for that particular purpose

### Environment

- VerneMQ Version: 1.8.0
- OS: macOS, linux

**Comment by larshesel** [comment]
Hi, thanks for the proposal. Can you describe the use case a bit more in detail - why is this feature necessary? 

> When a client that use `clean_session=true` have gone, all its subscriptions get removed. The issue that there is no way to find out about that action from within a plugin.

Changing the meaning of `on_subscribe/on_subscribe_m5` would not be an option as this would likely break existing plugins.

I don't think we cannot commit to option 1 as this would mean that we'd have to stick to the behaviour described forever, and the behaviour of `vmq_subscriber_db:read/1` is internal.

So I think the best option, should we decide to build support this feature, would be via option 3, unless we can think of some alternative.

**Comment by manifest** [comment]
> Hi, thanks for the proposal. Can you describe the use case a bit more in detail - why is this feature necessary?

We're currently developing a platform where applications and end-users connected all together via MQTT. Some topics assigned to applications based on conventions. When an end-user subscibes to some topic of a particular application we interpret that as an intend of the end-user to consume application's resources.

In our case, that's essential to notify the application about the end-user's intend. We use on_subscribe callback of our VerneMQ plugin to send a message about the fact that subscription was created to the application. We also need a way to send a message about the fact that subscription was deleted, including the cases when the end-user closed its session or its session was closed by the broker, and that's where the issue rise.

In other words, we need a way to send a notification message when a subscription is created and when its deleted. The issue is that at the moment there is no way to find out about the subscription deletion from within a plugin.

**Comment by manifest** [comment]
Just to generalize the description of the case and to clarify a reasoning behind proposed feature a little bit.

At the moment, the only way to send a notification that a subscription was deleted is to store information about the subscription in a local state (such as ets) of a broker node at the moment when the subscription was created. Then, to implement "on_client_gone" callback from where the notification would be sent if information about the subscription were presented in the local state. That means, to subscribe some client, it's **required to know the exact node where that client connected**.

The proposed feature will allow for much simpler approach. With callback informing us about deleting of a subscription, we don't need any local state such as one described above, because the state's only purpose is to preserve (or duplicate) information about the subscription. That means, we **can ask any broker node** to subscribe a client.

**Comment by manifest** [comment]
> So I think the best option, should we decide to build support this feature, would be via option 3, unless we can think of some alternative.

That might be some generalized callback that catch different events of the node including subscription creation and deletion.

**Comment by larshesel** [comment]
Thanks for the additional info. I can't give you any feedback right now as we won't be able to discuss this internally until the second week of October due to vacation.

**Comment by manifest** [comment]
@larshesel Thanks for letting me know. We'll adjust our development according to that.

**Comment by manifest** [comment]
@larshesel Have you had a chance to discuss the issue?

At the moment, we developed a work around for the issue that utilizes the ondeliver callback (it seems as the only way to access subscription data stored locally on a node). That solution is quite inefficient.

**Comment by larshesel** [comment]
Hi @manifest - we haven't had a chance to discuss this yet - we had to postpone the discussion a bit. /cc @dergraf

**Comment by manifest** [comment]
Just a humble attempt to bring attention to this issue :-) Are there any updates?

**Comment by ioolkos** [comment]
@manifest thanks.
I do not fully understand your idea of "some generalized callback that catch different events of the node including subscription creation and deletion". But let's discuss your ideas on that point. (if this is something adding general value, I'd probably be open to a PR, but let's first discuss)

**Comment by manifest** [comment]
In our case we need a way to create and delete subscriptions for particular clients, and get informed about such events on other clients.

We didn't find anything like that in the broker admin API, so we implemented it within a plugin: we now can send MQTT message (to some sort of "system" topic) with request to subscribe a client to the topic specified in the message. When subscription is created or deleted we send another MQTT message (or notification) to inform other clients about that operation. The current implementation works, but it is inefficient.

**A general value for this issue would be to allow plugin developers work with subscriptions from within a plugin and get notified about subscription changes**.

The proposal is to add a new hook that will be called by the broker when subscription is created or deleted, so plugin developers would be able to implement any logic they want for such events.

@ioolkos

**Comment by ioolkos** [comment]
@manifest thanks, I see. 
I think this would be named something like  `on_topic_created`, and `on_topic_deleted`, but called for every ClientID individually. I'm not yet 100% sure of the implications, where to hook this in, whether this has performance implications (don't really think so) etc.
But certainly willing to review PRs for this.

**Comment by manifest** [comment]
@ioolkos I got it. Thanks.
